json/src/value.rs

```diff
@@ -468,10 +468,14 @@ struct WriterFormatter<'a, 'b: 'a> {
 
 impl<'a, 'b> io::Write for WriterFormatter<'a, 'b> {
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        match self.inner.write_str(str::from_utf8(buf).unwrap()) {
-            Ok(_) => Ok(buf.len()),
-            Err(_) => Err(io::Error::last_os_error()),
+        fn io_error<E>(_: E) -> io::Error {
+            // Value does not matter because fmt::Debug and fmt::Display impls
+            // below just map it to fmt::Error
+            io::Error::new(io::ErrorKind::Other, "fmt error")
         }
+        let s = try!(str::from_utf8(buf).map_err(io_error));
```

Comment:
In `.or(io_error())` the argument is evaluated even if there is no error. And `io::Error::new` is [actually pretty expensive](https://github.com/rust-lang/rust/blob/de78655bca47cac8e783dbb563e7e5c25c1fae40/src/libstd/io/error.rs#L191-L204) - two memory allocations, one on the `into()` call that makes a Box, and another explicit `Box::new`. So this function would end up doing 4 useless memory allocations on every call to `write`... which may itself by called many many times as a value is being serialized.